quantecon/game_theory/polymatrix_game.py

```diff
@@ -0,0 +1,243 @@
+    Approximates the payoffs to a player when they play
+    an action with values at the actions of other players
+    that try to sum to the payoff.
+    Precise when the game can be represented with a polymatrix.
+
+    Parameters
+    ----------
+    nfg : NormalFormGame
+        The game.
+
+    my_player_number : int
+        The number of the player making the action.
+
+    my_action_number : int
+        The number of our player's action.
+
+    is_polymatrix : bool, optional
+        Is the game represented by this normal form
+        actually a polymatrix game. Defaults to False.
+
+    Returns
+    -------
+    Mapping[tuple[int, int], float]
+        Dictionary giving a (approximate) component
+        of the payoff at each other player and their action.
+    """
+    action_combinations = product(*(
+        [
+            range(nfg.nums_actions[p]) if p != my_player_number
+            else [my_action_number]
+            for p in range(nfg.N)
+        ]
+    ))
+    my_player: Player = nfg.players[my_player_number]
+    my_payoffs = my_player.payoff_array
+    hh_actions_and_payoffs = np.vstack([
+        np.hstack(
+            [
+                np.eye(nfg.nums_actions[p])[action_combination[p]]
+                for p in range(nfg.N) if p != my_player_number
+            ] + [
+                my_payoffs[
+                    action_combination[my_player_number:]
+                    + action_combination[:my_player_number]
+                ]
+            ]
+        )
+        for action_combination in action_combinations
+    ])
+    hh_actions = hh_actions_and_payoffs[:, :-1]
+    combined_payoffs = hh_actions_and_payoffs[:, -1]
+
+    # different ways to solve the simultaneous equations
+    if is_polymatrix:
+        # this does not go much faster and
+        # could also be used for games with no actual polymatrix
+        hh_payoffs_array, residuals, _, _ = np.linalg.lstsq(
+            hh_actions, combined_payoffs, rcond=None
+        )
+        assert np.allclose(
+            residuals, 0), "The game is not actually a polymatrix game."
+    else:
+        hh_payoffs_array = np.dot(
+            np.linalg.pinv(hh_actions), combined_payoffs)
+
+    payoff_labels = [
+        (p, a)
+        for p in range(nfg.N) if p != my_player_number
+        for a in range(nfg.nums_actions[p])
+    ]
+
+    payoffs = {
+        label: payoff
+        for label, payoff in zip(payoff_labels, hh_payoffs_array)}
+
+    return payoffs
+
+
+class PolymatrixGame:
+    """
+    Polymatrix Game.
+
+    `polymatrix[(a, b)]` is a 2D matrix, the payoff matrix of `a`
+    in the bimatrix game between `a` and `b`;
+    player number `a` is the row player
+    and player number `b` is the column player.
+
+    Attributes
+    ----------
+    number_of_players : int
+        Number of players.
+
+    nums_actions : Sequence[int]
+        The number of actions available to each player.
+
+    polymatrix : Mapping[ tuple[int, int], Matrix ]
```

Comment:
Is `self.polymatrix` not restricted to a `dict` with `np.array` as values?

Reply:
I am not sure on best practice with type hints regarding the numpy parts. So I made a type alias `Matrix` and left it as `any` as a placeholder but it should be a 2D numpy array (I could not find a neat way to type hint this). When I return one from `hh_player_payoffs`, I should specify it is a dict; however if someone wants to construct a polymatrix game using some other mapping, I don't think I need to stop them.